axlearn/common/attention.py

```diff
@@ -3934,6 +3935,33 @@ def forward(
 _SavePattern = Union[str, re.Pattern, None]
 
 
+def save_only_these_regex_patterns(*regex_patterns_to_save):
```

Comment:
How is this different from https://github.com/apple/axlearn/blob/92205bc8928adaef14b980f7e4dd4936ee953e2f/axlearn/common/attention.py#L3933-L3956?

Reply:
+1, you should probably move 
[axlearn/axlearn/common/attention.py](https://github.com/apple/axlearn/blob/92205bc8928adaef14b980f7e4dd4936ee953e2f/axlearn/common/attention.py#L3933-L3956)

into https://github.com/apple/axlearn/blob/92205bc8928adaef14b980f7e4dd4936ee953e2f/axlearn/common/utils.py#L129 and make it part of extended_remat_policies, if you need it outside of attention.py.

Reply:
I agree on reusing this policy, will also modify this policy to ingest multiple regex patterns as a sequence. I feel It is hard to put all remat names in just a single regex pattern.

Reply:
Can you elaborate on why it's hard to put all names into a single regex? You can use "|" to separate patterns and use "()" to make things more readable.

Reply:
Yes, by hard I meant readable. It is this
regex_patterns_to_save=[
    r"TransformerAttentionLayer\.residual_add",
    r"\.?(k|q|v)_proj$",
    r"\.?linear1_0",
    r"\.?linear1_1",
    r"TransformerFeedForwardLayer\.mlp_residual",
]
vs 
r"TransformerAttentionLayer\.residual_add|\.?(k|q|v)_proj$|\.?linear1_[01]|TransformerFeedForwardLayer\.mlp_residual"

Second one keeps the changes minimal, but the first may be a little cleaner. let me know the decision on this, I am okay with both

Reply:
I think we should reuse the API of `_save_and_offload_only_these_names_regex ` and avoid changing it since it appears in golden configs. Also cc @ruomingp for some input.

Reply:
`r"TransformerAttentionLayer\.residual_add|\.?(k|q|v)_proj$|\.?linear1_[01]|TransformerFeedForwardLayer\.mlp_residual"  `, regarding this part, is it possible to make it a few substrings and then concat them and make one, so that it is more readable while not creating a new API?

Reply:
I made updates to the PR to use `_save_and_offload_only_these_names_regex` as we discussed. The changes look clean. I put each different remat regex patterns neatly into different lines.
```
  names_which_can_be_saved=r"(TransformerAttentionLayer\.residual_add"  # pylint: disable=C0301
  "|.*\.?(k|q|v)_proj"
  "|.*\.?linear1_[01]"
  "|TransformerFeedForwardLayer\.mlp_residual)",
```